snuba/schemas.py

```diff
@@ -217,42 +183,149 @@ def get_time_series_query_schema_properties(default_granularity: int, default_wi
     }
 }
 
-EVENTS_QUERY_SCHEMA = {
-    **copy.deepcopy(GENERIC_QUERY_SCHEMA),
+PERFORMANCE_EXTENSION_SCHEMA = {
+    'type': 'object',
+    'properties': {
+        # Never add FINAL to queries, enable sampling
+        'turbo': {
+            'type': 'boolean',
+            'default': False,
+        },
+        # Force queries to hit the first shard replica, ensuring the query
+        # sees data that was written before the query. This burdens the
+        # first replica, so should only be used when absolutely necessary.
+        'consistent': {
+            'type': 'boolean',
+            'default': False,
+        },
+        'debug': {
+            'type': 'boolean',
+            'default': False,
+        },
+    },
+    'additionalProperties': False,
+}
+
+PROJECT_EXTENSION_SCHEMA = {
+    'type': 'object',
+    'properties': {
+        'project': {
+            'anyOf': [
+                {'type': 'number'},
+                {
+                    'type': 'array',
+                    'items': {'type': 'number'},
+                    'minItems': 1,
+                },
+            ]
+        },
+    },
     # Need to select down to the project level for customer isolation and performance
     'required': ['project'],
+    'additionalProperties': False,
 }
 
-EVENTS_QUERY_SCHEMA['properties'].update({
-    **get_time_series_query_schema_properties(
+
+def get_time_series_extension_properties(default_granularity: int, default_window: timedelta):
+    return {
+        'type': 'object',
+        'properties': {
+            'from_date': {
+                'type': 'string',
+                'format': 'date-time',
+                'default': lambda: (datetime.utcnow().replace(microsecond=0) - default_window).isoformat()
+            },
+            'to_date': {
+                'type': 'string',
+                'format': 'date-time',
+                'default': lambda: datetime.utcnow().replace(microsecond=0).isoformat()
+            },
+            'granularity': {
+                'type': 'number',
+                'default': default_granularity,
+            },
+        },
+        'additionalProperties': False,
+    }
+
+
+@dataclass
+class Request:
```

Comment:
frozen? (provided it allows you to add an additional field in __post_init__ (I never tried)

Reply:
This will throw `FrozenInstanceError` but I can make `body` a property instead, the `ChainMap` should be pretty lightweight (compared to a dictionary copy) and is only used once right now anyway.

Reply:
Did that with f85b70e2e8b5c1087ee0563cb583e70d9adc7963.